src/databricks/labs/ucx/source_code/redash.py

```diff
@@ -1,24 +1,171 @@
+            return
+        # create the backup folder for queries if it does not exist
+        try:
+            self._ws.workspace.get_status(self._backup_path)
+        except ResourceDoesNotExist:
+            self._ws.workspace.mkdirs(self._backup_path)
+        backup_query = self._ws.queries.create(
             data_source_id=query.data_source_id,
             description=query.description,
-            name=query.name,
-            options=query.options,
+            name=str(query.name) + "_original",
+            options=query.options.as_dict() if query.options is not None else None,
+            parent=self._backup_path,
             query=query.query,
             run_as_role=query.run_as_role,
+            tags=[self.BACKUP_TAG],
         )
+        from_table = FromTable(self._index, self._get_session_state(query))
+        new_query = from_table.apply(query.query)
+        try:
+            self._ws.queries.update(
+                query.id,
+                query=new_query,
+                tags=self._get_migrated_tags(query.tags, [f'backup:{backup_query.id}']),
+            )
+        except DatabricksError:
+            logger.error(f"Cannot upgrade {query.name}")
+            return
+
+    @staticmethod
+    def _get_session_state(query: Query) -> CurrentSessionState:
+        session_state = CurrentSessionState()
+        if query.options is None:
+            return session_state
+        if query.options.catalog:
+            session_state = replace(session_state, catalog=query.options.catalog)
+        if query.options.schema:
+            session_state = replace(session_state, schema=query.options.schema)
+        return session_state
+
+    def _revert_query(self, query: Query):
+        assert query.id is not None
+        assert query.query is not None
+        if query.tags is None:
+            return
+        # find the backup query
+        is_migrated = False
+        backup_id = None
+        for tag in query.tags:
+            if tag == self.MIGRATED_TAG:
+                is_migrated = True
+                continue
+            if tag.startswith("backup:"):
+                backup_id = tag.split(":")[1]
+                continue
+
+        if not is_migrated:
+            logger.debug(f"Query {query.name} was not migrated by UCX")
+            return
+
+        if backup_id is None:
+            logger.debug(f"Cannot find backup query for query {query.name}")
+            return
+        try:
+            original_query = self._ws.queries.get(backup_id)
+            self._ws.queries.update(query.id, query=original_query.query, tags=self._get_original_tags(query.tags))
+            self._ws.queries.delete(backup_id)
+        except DatabricksError:
+            logger.error(f"Cannot restore {query.name} from backup query {backup_id}")
+            return
+
+    def _get_migrated_tags(self, tags: list[str] | None, new_tags: list[str] | None = None) -> list[str]:
```

Comment:
both `tags` and `new_tags` look redundant, could be only `tags`

Reply:
fixed